extensions/wikia/Recirculation/js/helpers/LiftigniterHelper.js

```diff
@@ -5,95 +5,116 @@ define('ext.wikia.recirculation.helpers.liftigniter', [
 ], function ($, w, thumbnailer) {
 	'use strict';
 
-	var helper = function (config) {
-		var defaults = {
-				max: 5,
-				width: 320,
-				height: 180,
-				flush: false
-			},
-			options = $.extend({}, defaults, config);
-
-		function loadData() {
-			var deferred = $.Deferred(),
-				registerOptions = {
-					max: options.max * 2, // We want to load twice as many because we filter based on thumbnails
-					widget: options.widget,
-					callback: function (response) {
-						deferred.resolve(formatData(response));
+	var willFetch = {},
+		fetchQueue = [],
+		fetching = false;
+
+	var prepare = function (options) {
+
+		var deferred = $.Deferred(),
+			registerOptions = {
+				max: options.max,
+				widget: options.widget,
+				callback: function (response) {
+					deferred.resolve(response);
+					fetching = false;
+					if(fetchQueue.length) {
+						// setTimeout is necessary because LI need to wait for LI to remove already registered modules
+						// LI clear register modules just after running all callbacks
+						setTimeout(function () {
+							fetch(fetchQueue.shift());
+						});
 					}
-				};
-
-			if (!w.$p) {
-				return deferred.reject('Liftigniter library not found').promise();
-			}
-
-			if (options.opts) {
-				registerOptions.opts = options.opts;
-			}
-
-			// Callback renders and injects results into the placeholder.
-			w.$p('register', registerOptions);
-
-			if (options.flush) {
-				w.$p('fetch');
-			}
+				}
+			};
 
-			return deferred.promise();
+		if (!w.$p) {
+			return deferred.reject('Liftigniter library not found').promise();
 		}
 
-		function formatData(data) {
-			var items = [],
-				title = '';
-
-			if (options.title) {
-				title = options.title;
-			}
-
-			$.each(data.items, function (index, item) {
-				item.isWiki = false;
-
-				if (items.length < options.max && item.thumbnail) {
-					item.source = options.source;
-					item.meta = options.widget;
-					item.index = index;
-
-					if (item.source === 'wiki') {
-						item.isWiki = true;
-					}
-					if (thumbnailer.isThumbUrl(item.thumbnail)) {
-						item.thumbnail = thumbnailer.getThumbURL(item.thumbnail, 'image', options.width, options.height);
-					}
+		if (options.opts) {
+			registerOptions.opts = options.opts;
+		}
 
-					items.push(item);
-				}
-			});
+		var module = {
+			promise: deferred,
+			options: registerOptions
+		};
 
-			return {
-				title: title,
-				items: items
-			};
+		if(willFetch[options.modelName]) {
+			willFetch[options.modelName].push(module)
+		} else {
+			willFetch[options.modelName] = [module];
 		}
 
-		function setupTracking() {
-			var elements = $('.recirculation-unit .item[data-meta="' + options.widget + '"]').get(),
-				trackOptions = {
-					elements: elements,
-					name: options.widget,
-					source: 'LI'
-				};
+		return deferred.promise();
+		// TODO thumbnail sizes, setup tracking
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)